### PR TITLE
Refactored code to use cLuaState

### PR DESCRIPTION
**Pull request by SamJBarney**
* Fixes https://github.com/cuberite/cuberite/issues/2377, as requested by @madmaxoft

### CODE REVIEW

File: src/Bindings/DeprecatedBindings.cpp [hunk 1]
```diff
@@ -293,20 +299,32 @@ static int tolua_cWorld_SetSignLines(lua_State * tolua_S)
...
-		int BlockX          = static_cast<int>(tolua_tonumber(LuaState, 2, 0));
-		int BlockY          = static_cast<int>(tolua_tonumber(LuaState, 3, 0));
-		int BlockZ          = static_cast<int>(tolua_tonumber(LuaState, 4, 0));
-		const AString Line1 = tolua_tocppstring(LuaState, 5, 0);
-		const AString Line2 = tolua_tocppstring(LuaState, 6, 0);
-		const AString Line3 = tolua_tocppstring(LuaState, 7, 0);
-		const AString Line4 = tolua_tocppstring(LuaState, 8, 0);
+		cWorld * self = nullptr;
+		cPlayer * Player = nullptr;
+		int BlockX = 0;
```
Comment (NiLSPACE) [inline src/Bindings/DeprecatedBindings.cpp]: Can't these be one line?
```cpp
int BlockX, BlockY, BlockZ
```
Or do we not accept that?

Comment (SamJBarney) [inline src/Bindings/DeprecatedBindings.cpp]: I honestly don't know what the coding style is on that, but I do know they need to be initialized to 0.

Comment (SphinxC0re) [inline src/Bindings/DeprecatedBindings.cpp]: I personally find this more readable

File: src/Bindings/DeprecatedBindings.cpp [hunk 2]
```diff
@@ -293,20 +299,32 @@ static int tolua_cWorld_SetSignLines(lua_State * tolua_S)
...
+
+		LuaState.GetUserType(1, self, "cWorld");
+		LuaState.GetUserType(9, Player, "cPlayer");
+		LuaState.GetStackValue(2, BlockX);
+		LuaState.GetStackValue(3, BlockY);
+		LuaState.GetStackValue(4, BlockZ);
+		LuaState.GetStackValue(5, Line1);
+		LuaState.GetStackValue(6, Line2);
+		LuaState.GetStackValue(7, Line3);
+		LuaState.GetStackValue(8, Line4);
```
Comment (NiLSPACE) [inline src/Bindings/DeprecatedBindings.cpp]: Perhaps we can create a cLuaState::GetStackValue**s** function
It would look like this:
```cpp
cLuaState::GetStackValues(int a_MinStackPos, int a_MaxStackPos, ...)
{
   va_list argList;
   va_start(argList, a_MaxStackPos);
   for (int i = a_MinStackPos; i <= a_MaxStackPos; i++)
   {
      // Is it possible to have a variable number of parameters as references? If so, then set them here.
   }
   va_end(argList)
}
```

Comment (SamJBarney) [inline src/Bindings/DeprecatedBindings.cpp]: Either a var arg, or a variadic template

Comment (SphinxC0re) [inline src/Bindings/DeprecatedBindings.cpp]: That's a different issue. We have to focus on, not cascade through problems

Comment (NiLSPACE) [inline src/Bindings/DeprecatedBindings.cpp]: Well, otherwise it would require another refactor.

Comment (SphinxC0re) [inline src/Bindings/DeprecatedBindings.cpp]: Yeah but that's waht I would prefer: Doing it layer by layer. Refactory should take a long time and many steps

**Comment by SamJBarney** [comment]
Well, other than the debate about style and the creation of cLuaState::GetStackValues, does this look good for squash?

**Comment by NiLSPACE** [comment]
@madmaxoft probably wants to take a look as well.

File: src/Bindings/DeprecatedBindings.cpp (hunk 2 above)
Comment (madmaxoft) [inline src/Bindings/DeprecatedBindings.cpp]: There already is a `cLuaState::GetStackValues()` function. Its first parameter is the staring index, and then it's just a list of variables which should be read. It's templated so that any number and any type of parameters are supported; it calls `GetStackValue` internally.

File: src/Bindings/DeprecatedBindings.cpp (hunk 1 above)
Comment (madmaxoft) [inline src/Bindings/DeprecatedBindings.cpp]: With initialization, use one line per variable. Without initialization (like the Line1, Line2 etc below), I prefer to merge them into a single line. But that's just personal opinion.

File: src/Bindings/LuaState.cpp [hunk 3]
```diff
@@ -845,6 +843,21 @@ bool cLuaState::GetStackValue(int a_StackPos, float & a_ReturnedVal)
...
+bool cLuaState::GetStackValue(int a_StackPos, int & a_Value)
+{
+	double tmp;
+	if (GetStackValue(a_StackPos, tmp))
+	{
+		a_Value = FloorC(tmp);
+		return true;
+	}
+	return false;
+}
```
Comment (madmaxoft) [inline src/Bindings/LuaState.cpp]: This is not needed at all, there's already a templated function handling `int`s:
https://github.com/cuberite/cuberite/blob/master/src/Bindings/LuaState.h#L268-L287

File: src/Bindings/LuaState.h [hunk 4]
```diff
@@ -263,6 +266,19 @@ class cLuaState
...
+	// If not, a_UserType remains unchanged.
+	template <typename T>
+	bool GetUserType(int a_StackPos, T * & a_UserType, const char * a_Type)
+	{
+		if (CheckParamUserType(a_StackPos, a_Type))
+		{
+			a_UserType = static_cast<T *>(tolua_tousertype(m_LuaState, a_StackPos, nullptr));
+		}
+		return false;
+	}
```
Comment (madmaxoft) [inline src/Bindings/LuaState.h]: This is not needed. Client code should first check all parameters, and then use `GetStackValue()` or `GetStackValues()`.

I'm considering a rewrite of the `CheckParamXYZ` handling so that it would better support overloaded functions, and for that I need the current usage idiom.

File: src/Bindings/ManualBindings.cpp [hunk 5]
```diff
@@ -1313,41 +1313,48 @@ static int tolua_SetObjectCallback(lua_State * tolua_S)
...
+		else
 		{
-			luaL_unref(tolua_S, LUA_REGISTRYINDEX, Reference);
+			LOGWARNING("cPluginLua:AddWebTab: invalid function reference in 2nd argument (Title: \"%s\")", Title.c_str());
 		}
 	}
 	else
 	{
-		LOGWARNING("cPluginLua:AddWebTab: invalid function reference in 2nd argument (Title: \"%s\")", Title.c_str());
+		LOGWARNING("cPluginLua::AddWebTab: invalid self as first argument");
```
Comment (madmaxoft) [inline src/Bindings/ManualBindings.cpp]: Lua uses a single colon for class-function calls, so we report it that way.

File: src/Bindings/LuaState.h (hunk 4 above)
Comment (SamJBarney) [inline src/Bindings/LuaState.h]: @madmaxoft I think that GetUserType is necessary, because there are no GetStackValue methods that would handle user types correctly.

How are you planning on rewriting the GetParamXYZ handling? Maybe I can update my code based on that.

Comment (madmaxoft) [inline src/Bindings/LuaState.h]: The usertype-based `GetStackValue` functions are generated into the $/src/Bindings/LuaState_Declaration.inc and _implementation.cpp files automatically during the build process.

**Comment by SamJBarney** [comment]
Well, at least I know more about how the bindings are being generated. Though they shouldn't need to be generated at buildtime with a proper template method.

**Comment by madmaxoft** [comment]
Templates can't turn class names into strings, which is needed for the custom types.

File: src/Bindings/DeprecatedBindings.cpp [hunk 6]
```diff
@@ -293,20 +299,32 @@ static int tolua_cWorld_SetSignLines(lua_State * tolua_S)
 		!tolua_iscppstring(LuaState, 8, 0, &tolua_err) ||
 		!tolua_isusertype (LuaState, 9, "cPlayer", 1, &tolua_err) ||
 		!tolua_isnoobj    (LuaState, 10, &tolua_err)
+		*/
 		)
 		goto tolua_lerror;
```
Comment (madmaxoft) [inline src/Bindings/DeprecatedBindings.cpp]: Return immediately, the `CheckParamXYZ()` functions already do output the error message so there's no need to do anything more.

File: src/Bindings/DeprecatedBindings.cpp [hunk 7]
```diff
@@ -293,20 +299,32 @@ static int tolua_cWorld_SetSignLines(lua_State * tolua_S)
...
+		int BlockX = 0;
+		int BlockY = 0;
+		int BlockZ = 0;
+		AString Line1;
+		AString Line2;
+		AString Line3;
+		AString Line4;
+
+		LuaState.GetStackValue(1, self);
+		LuaState.GetStackValue(9, Player);
```
Comment (madmaxoft) [inline src/Bindings/DeprecatedBindings.cpp]: Sort by stack index and use `GetStackValues`

File: src/Bindings/LuaState.h [hunk 8]
```diff
@@ -35,6 +35,9 @@ extern "C"
 #include "PluginManager.h"
 #include "LuaState_Typedefs.inc"
 
+#undef TOLUA_TEMPLATE_BIND
+#include "tolua++/include/tolua++.h"
+
```
Comment (madmaxoft) [inline src/Bindings/LuaState.h]: Revert this change, move the inclusion back into the cpp file. We don't want to include the tolua library into each compilation unit including LuaState.h

File: src/Bindings/ManualBindings.cpp [hunk 9]
```diff
@@ -1313,41 +1313,48 @@ static int tolua_SetObjectCallback(lua_State * tolua_S)
...
 {
-	cPluginLua * self = reinterpret_cast<cPluginLua *>(tolua_tousertype(tolua_S, 1, nullptr));
+	cLuaState LuaState(tolua_S);
+	cPluginLua * self = nullptr;
 
-	tolua_Error tolua_err;
-	tolua_err.array = 0;
-	tolua_err.index = 3;
-	tolua_err.type = "function";
+	if (LuaState.GetStackValue(1, self))
```
Comment (madmaxoft) [inline src/Bindings/ManualBindings.cpp]: Use opposite condition and return out of the function, to avoid nesting too deep.

**Comment by SamJBarney** [comment]
@madmaxoft any more changes?

File: src/Bindings/DeprecatedBindings.cpp [hunk 10]
```diff
@@ -281,32 +281,32 @@ static int tolua_cWorld_SetSignLines(lua_State * tolua_S)
...
+		int BlockZ = 0;
+		AString Line1;
+		AString Line2;
+		AString Line3;
+		AString Line4;
+
+		LuaState.GetStackValue(1, self);
+		LuaState.GetStackValues(2, BlockX, BlockY, BlockZ);
+		LuaState.GetStackValues(5, Line1, Line2, Line3, Line4);
+		LuaState.GetStackValue(9, Player);
```
Comment (madmaxoft) [inline src/Bindings/DeprecatedBindings.cpp]: Why not just a single call to `GetStackValues()`? It does support variable length of varying parameters, you know :)
https://github.com/cuberite/cuberite/blob/master/src/Bindings/ManualBindings_World.cpp#L41

Comment (SamJBarney) [inline src/Bindings/DeprecatedBindings.cpp]: I think it is just better semantically to get similar stack values with a single method call than all stack values with a single method call.

Plus, GetStackValues causes a bit of binary bloat. Not a lot, but some. And by grouping like I am, I can hopefully cut down on it a little bit.

Comment (madmaxoft) [inline src/Bindings/DeprecatedBindings.cpp]: I think it's much better to use just one call, because it makes it easy to see the function's parameters, from the API standpoint this is excellent.

I don't think there's much difference between the amount of code generated for the two.

Comment (worktycho) [inline src/Bindings/DeprecatedBindings.cpp]: There's some code bloat, though not much but it would be far more effective to optimize that explicitly if necessary, rather than on an ad-hoc basis.

File: src/Bindings/ManualBindings.cpp [hunk 11]
```diff
@@ -1348,6 +1352,7 @@ static int tolua_cPluginLua_AddWebTab(lua_State * tolua_S)
 		LOGWARNING("cPluginLua:AddWebTab: invalid function reference in 2nd argument (Title: \"%s\")", Title.c_str());
 	}
 
+
```
Comment (madmaxoft) [inline src/Bindings/ManualBindings.cpp]: Get rid of this extra line when rebasing.

**Comment by madmaxoft** [comment]
Looks good, rebase and squash (plus edit out the extra newline)

**Comment by SamJBarney** [comment]
Squashed

**Comment by madmaxoft** [comment]
I merged this manually, because I didn't like the commit message :P

It would have been much easier if you developed on a branch inside the main repo.
